### ISSUE TITLE
Remove old code generation methods of stream methods in service interfaces

### ISSUE THREAD

**Issue by stevenroose** [grpc/grpc-go#1894]
For a bi-directional stream, I get the following generated code:
```go
type Btcd_SubscribeTransactionsServer interface {
	Send(*MempoolTransaction) error
	Recv() (*SubscribeTransactionsRequest, error)
	grpc.ServerStream
}
```

There is absolutely no documentation about the behavior of these methods. They should either be documented in the code, or have a link to a page with documentation about them.

The "Go Basics" documentation merely gives the following example code:
```go
func (s *routeGuideServer) RouteChat(stream pb.RouteGuide_RouteChatServer) error {
	for {
		in, err := stream.Recv()
		if err == io.EOF {
			return nil
		}
		if err != nil {
			return err
		}
		key := serialize(in.Location)
                ... // look for notes to be sent to client
		for _, note := range s.routeNotes[key] {
			if err := stream.Send(note); err != nil {
				return err
			}
		}
	}
}
```

It seems that it's impossible to get a channel for incoming messages. But it's also not clear if `Recv` is a blocking call. Does `io.EOF` mean that the client closed that channel or that there is currently nothing to be read?

**UPDATE:** 

Implementations for streams are modified to use generics https://github.com/grpc/grpc-go/pull/7057, which will cause the gRPC codegen to use prebuilt generic types to implement client and server stream objects, rather than generating new types and implementations for every RPC method. We need to make sure the stream interfaces using generics are well documented. Here are the high level tasks for completion

- [x] Add documenation for stream interfaces - https://github.com/grpc/grpc-go/pull/7470
- [ ] Delete the code to generate stream interfaces for client and server.
- [x] update the [grpc.io docs](https://grpc.io/docs/languages/go/generated-code/) to use generics instead of generated interfaces.

**Comment by dfawley** [comment]
`Recv` is blocking. io.EOF from `Recv` means the client called `CloseSend` (or `SendAndClose`).

There is a [manual](https://grpc.io/docs/reference/go/generated-code.html) for our generated code on grpc.io, but I agree we should have better godocs or pointers to the manual from the generated code.

**Comment by dfawley** [comment]
> So it's not really possible to do send and receive both from the same goroutine?

Right, this is not recommended (but it is technically possible if you control both server and client and can ensure they work together properly).  `Send()` is also blocking, so if the other side is not receiving, it will eventually run out of flow control and block until the other side does call `Recv()`.  This could lead to a deadlock if you aren't careful.

> It seems that I should exclude io.EOF in that small goroutine from closing the channel. So that the client can still receive notifications when it no longer wants to change it's subscription. Or explicitly document that the server closes the stream when the client does.

Either way seems reasonable for this use case.  I don't think there's much value in calling `CloseStream` from the client except when it's ready to end the stream, and then you can get the final status in case it's potentially interesting (which you can't get if you have to cancel the context to end the stream).

**Comment by srini100** [comment]
Using this bug to consolidate issues regarding documentation around streaming API. 
https://github.com/grpc/grpc-go/issues/1880
https://github.com/grpc/grpc-go/issues/1876

**Comment by dfawley** [comment]
@srini100, note that we've recently added some stream documentation here:

https://github.com/grpc/grpc-go/blob/master/Documentation/concurrency.md

Is that sufficient for all of these issues?

**Comment by stevenroose** [comment]
@dfawley I'd primarily like code documentation on the generated methods. So that I can read it (1) from godocs and (2) from going into the code with my IDE. Mostly the 2nd is where I was missing it. When you're implementing your server, you are using these interfaces. So it's normal that you want to go look at the documented behavior of them, right? Right now you get nothing.

**Comment by floatdrop** [comment]
There are still some things that could be clarified more about `Send()` method on stream: we receiving `io.EOF` errors after calling this method and only one possible answer can be found in mailing list – https://groups.google.com/d/msg/grpc-io/XcN4hA9HonI/F_UDiejTAwAJ

Can we clarify what errors and how they should be handled may `Send()` method return?

**Comment by purnesh42H** [comment]
@janardhanvissa recently the [implementations for streams are modified to use generics](https://github.com/grpc/grpc-go/pull/7057) which will cause the gRPC codegen to use prebuilt generic types to implement client and server stream objects, rather than generating new types and implementations for every RPC method

So, most of the above discussion is not relevant anymore. For now, you can add inline comments to stream interfaces's methods following the documentation here https://grpc.io/docs/languages/go/generated-code/

@dfawley to confirm if that is enough for this task

**Comment by dfawley** [comment]
+1.  The medium-term plan is now to completely delete these generated interfaces in favor of the generics (they are still optionally generated for now).  We should make sure the generics are well documented and make sure the generated code has enough information to help users understand them.  We should also update the grpc.io doc @purnesh42H pointed to as needed.  Another thing I noticed yesterday is that it doesn't mention the embedding of the `Unimplemented<Service>Server` struct inside implementations of the service.  That should ideally be added there as well.

**Comment by janardhanvissa** [comment]
@purnesh42H, I have added the behavior of the function "RouteChat" and generated the protobuf files. I have raised a PR for the changes that made. Please review the PR and let us know if any changes are required for the ticket. Please find the PR below. 

https://github.com/grpc/grpc-go/pull/7441

**Comment by purnesh42H** [comment]
@janardhanvissa the PR is not addressing the requirements of this task. As mentioned above, the implementations for streams are modified to use generics https://github.com/grpc/grpc-go/pull/7057, so we need to document the stream interfaces in generics. Here are the high level tasks that needs to be done.
1. Add the inline comments for interfaces in https://github.com/grpc/grpc-go/blob/566aad1ffd25190a01c160e1a7f4ca645b4896dc/stream_interfaces.go#L24
Use the explainations from here https://grpc.io/docs/languages/go/generated-code/

2. delete these generated interfaces (if there are any)

3. update the grpc.io docs to use generics instead of generated interfaces.

Let's modify the PR you sent for 1) and then we will move to 2) and 3)

**Comment by dfawley** [comment]
> 2. delete these generated interfaces (if there are any)

FWIW this is not correct.  Generated code should never be directly modified.  It should be regenerated when the generator or source files change, which our vet-proto checker will fail if we don't  do that correctly.

**Comment by janardhanvissa** [comment]
@purnesh42H Added Inline comments for all the 6 interfaces in stream_interfaces.go. Please find the PR below.

https://github.com/grpc/grpc-go/pull/7470

**Comment by dfawley** [comment]
@purnesh42H - you reopened this -- what still needs to be done here?

**Comment by purnesh42H** [comment]
> @purnesh42H - you reopened this -- what still needs to be done here?
> 
> yeah, don't we still have to do these 2?
> 
> * Delete the code to generate stream interfaces for client and server like here https://github.com/grpc/grpc-go/blob/master/cmd/protoc-gen-go-grpc/grpc.go#L337-L346 ?
> * update the grpc.io docs to use generics instead of generated interfaces. https://grpc.io/docs/languages/go/generated-code/ is not applicable anymore

yeah, don't we still have to do these 2?

- Delete the code to generate stream interfaces for client and server like here https://github.com/grpc/grpc-go/blob/master/cmd/protoc-gen-go-grpc/grpc.go#L337-L346 ?
- update the grpc.io docs to use generics instead of generated interfaces. https://grpc.io/docs/languages/go/generated-code/ is not applicable anymore

**Comment by purnesh42H** [comment]
> Sure. Maybe we can update the title of the issue to include these things, too? We added documentation to the generic interfaces, so from the title it seems like we're done here.

Updated

**Comment by eshitachandwani** [comment]
We will need to update the documentation to remove the part that says that we can go back to old behavior by changing the flag, when we remove the flag and make the use of stream generics default

**Comment by eshitachandwani** [comment]
Quick summary:
There were 3 tasks here :

- [X] Add documentation for stream interfaces - grpc: add docs for generic stream interfaces  #7470

- [ ] Delete the code to generate stream interfaces for client and server.

- [X] update the [grpc.io docs](https://grpc.io/docs/languages/go/generated-code/) to use generics instead of generated interfaces. - https://github.com/grpc/grpc.io/pull/1365

Task 1 and 3 are done.
For task 2 , after removing the old code to generate stream interfaces, it caused some issues in google3 and so we had to revert the change.
 - The PR to remove the code: https://github.com/grpc/grpc-go/pull/7654
 - Reverted PR : https://github.com/grpc/grpc-go/pull/7703

**Comment by eshitachandwani** [comment]
Changing the title to reflect the same.

**Comment by eshitachandwani** [comment]
The flag in internal google3 is still default to false. We need to turn it to true. We can remove the old code generation methods after the flag has been true in google3 for some time.

**Comment by eshitachandwani** [comment]
The flag internally has been turned to true , we need to wait atleast a month to make sure it does not break anyone.
Blocked  at least until : 22/02/26
